components/core/src/os/net/mod.rs

```diff
@@ -22,3 +22,71 @@ mod imp;
 mod imp;
 
 pub use self::imp::*;
+
+extern crate dns_lookup;
+
+use dns_lookup::{getaddrinfo,
+                 AddrInfoHints};
+use std::io;
+
+// lookup_fqdn returns a vector of fqdn that resolves the provided hostname.
+//
+// Since the underlying crate is platform agnostic, this function is as well,
+// we only implement a single function `ai_canonname()` that returns the right
+// hint flag for the running operating system.
+pub fn lookup_fqdn(hostname: &str) -> io::Result<Vec<String>> {
+    let hints = AddrInfoHints { flags: ai_canonname(),
+                                ..AddrInfoHints::default() };
+    let addrinfos =
+        getaddrinfo(Some(hostname), None, Some(hints))?.collect::<std::io::Result<Vec<_>>>()?;
+
+    let canonnames = addrinfos.into_iter()
+                              .filter_map(|info| info.canonname)
+                              .collect();
```

Comment:
Reading the [manpage for getaddrinfo](http://man7.org/linux/man-pages/man3/getaddrinfo.3.html) reveals this nugget:

> If hints.ai_flags includes the AI_CANONNAME flag, then the ai_canonname field of the first of the addrinfo structures in the returned list is set to point to the official name of the host.

This means that we just need to look at the first item in the list, grab the canonical name out of it, and return a `String` rather than a `Vec<String>`, so that should simplify a lot of stuff here.

Reply:
@christophermaier Would it be ok to return an `Option<String>` since that is what `canonname` is already returning?